MIPS/mip-zkapps.md

```diff
@@ -0,0 +1,960 @@
+discussions-to:
+status: Draft
+type: Standards Track
+category: Core
+created: 2023-02-14
+---
+
+## Abstract
+
+This MIP proposes adding programmable smart contracts, called zkApps, to the Mina protocol.
+
+## Motivation
+
+The zkApps protocol aims to add a smart contract layer to Mina which extends the
+potential of zero-knowledge cryptography by enabling the following
+characteristics while preserving the succinctness of Mina:
+
+
+  **General programmability** - able to execute and settle arbitrary programs,
+  not constrained to a particular VM model, but instead flexibly designed to
+  support any number of execution models, VM-like or otherwise
+
+  **Programmable privacy** - the privacy of both inputs to smart contracts and
+  their state can be programmed by their developers
+
+  **Constant in verification time** - individual transactions are executed, or
+  more accurately “proven” asynchronously off-chain and verified on chain in
+  time proportional to the account updates, independent of computational
+  complexity of any individual proof inside of each account update
+
+
+This MIP represents the first iteration of such a protocol.
+
+## Specification
+
+The following description summarizes the planned changes to the protocol. Not
+all technical details are described here.
+
+### Ledger account changes
+
+Recall that accounts are identified by a public key and token id pair.
+
+To accommodate zkApps, Mina ledger accounts have the following new field:
+
+- `token_symbol`: a string, up to six characters long, that can provide a name
+     for a `token_id` whose owner is the public key and token id in this
+     account. In the existing mainnet, only the default MINA `token_id` is
+     currently used. New `token_id`s are derived from the the addresses of the
+     smart contract that defines them.
+
+The following account field is removed:
+
+- `token_permissions`: this field was meant for non-default tokens, which were
+never introduced
+
+The `snapp` field in accounts, which was never instantiated, is
+replaced by the `zkapp` field, which is optional.
+
+The `zkapp` field is a record with the following fields:
+
+- `app_state` : an array of eight field elements, the current state of a zkApp
+associated with the account
+- `verification_key` : an optional key to be used when verifying transaction
+SNARKs for zkApps using this account; when it is missing, the account is a
+typical Mina account that supports payments.
+- `zkapp_version` : a nonnegative integer, the zkApps version that determined
+the `app_state`
+- `action_state` : an array of five field elements, each representing an
+"action" (see below) contained in a zkApp
+- `last_action_slot` : the global slot when the `action_state` was last updated
+- `proved_state` : a boolean, whether the application state has been set from a
+proof-driven zkapp transaction.
+- `zkapp_uri` : a string, of max size=255, either empty, or denoting a URI that
```

Comment:
Is this 255 fields? Is it too late, to consider just adding this on to appState? Then developers can decide if they want to use it for a zkapp_uri or to just have more data for their zkApp

Reply:
It's a string, of max length 255 characters.

The idea was to encode it as the poseidon hash of these characters inside the circuit / in the hash, so that zkApps can unpack it to the string if they want to, but being a single field element makes it lightweight in the transaction snark.

Reply:
Oh that's cool, okay nvm then!